src/librustc_typeck/check/method/probe.rs

```diff
@@ -602,8 +654,18 @@ impl<'a, 'gcx, 'tcx> ProbeContext<'a, 'gcx, 'tcx> {
 
         // Check whether `trait_def_id` defines a method with suitable name:
         let trait_items = self.tcx.trait_items(trait_def_id);
-        let maybe_item = trait_items.iter()
-            .find(|item| item.name() == self.item_name);
+        let maybe_item = match self.looking_for {
+            LookingFor::MethodName(item_name) => {
+                trait_items.iter()
+                           .find(|item| item.name() == item_name)
+            }
+            LookingFor::ReturnType(item_ty) => {
+                trait_items.iter()
+                           .find(|item| {
+                                self.matches_return_type(item, &item_ty)
+                            })
+            }
+        };
```

Comment:
You should've replaced the original code to a call to `impl_or_trait_item` - since they're equivalent ;).